builtin/logical/pki/path_manage_keys.go

```diff
@@ -0,0 +1,173 @@
+package pki
+
+import (
+	"context"
+	"strings"
+
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/certutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+func pathGenerateKey(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "keys/generate/(internal|exported)",
+
+		Fields: map[string]*framework.FieldSchema{
+			keyNameParam: {
+				Type:        framework.TypeString,
+				Description: "Optional name to be used for this key",
+			},
+			keyTypeParam: {
+				Type:        framework.TypeString,
+				Default:     "rsa",
+				Description: `Type of the secret key to generate`,
+			},
+			"key_bits": {
+				Type:        framework.TypeInt,
+				Default:     2048,
+				Description: `Type of the secret key to generate`,
+			},
+		},
+
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.CreateOperation: &framework.PathOperation{
+				Callback:                    b.pathGenerateKeyHandler,
+				ForwardPerformanceStandby:   true,
+				ForwardPerformanceSecondary: true,
+			},
+		},
+
+		HelpSynopsis:    pathGenerateKeyHelpSyn,
+		HelpDescription: pathGenerateKeyHelpDesc,
+	}
+}
+
+const (
+	pathGenerateKeyHelpSyn  = `Generate a new private key used for signing.`
+	pathGenerateKeyHelpDesc = `This endpoint will generate a new key pair of the specified type (internal, exported, or kms).`
+)
+
+func (b *backend) pathGenerateKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyName, err := getKeyName(ctx, req.Storage, data)
+	if err != nil { // Fail Immediately if Key Name is in Use, etc...
+		return nil, err
+	}
+	keyType := data.Get(keyTypeParam).(string)
+	keyBits := data.Get("key_bits").(int)
+
+	switch {
+	case strings.HasSuffix(req.Path, "/internal"):
+		// Internal key generation, stored in storage
+		keyBundle, err := certutil.GetKeyBundleFromKeyGenerator(keyType, keyBits, nil)
+		if err != nil {
+			return nil, err
+		}
+		privateKeyPemString, err := keyBundle.ToPrivateKeyPemString()
+		if err != nil {
+			return nil, err
+		}
+		key, _, err := importKey(ctx, req.Storage, privateKeyPemString, keyName)
```

Comment:
Again, not yours necessarily to fix, but if you want to, there's a trivial bug in importKey: we need to normalize whitespace:

```diff
diff --git a/builtin/logical/pki/storage.go b/builtin/logical/pki/storage.go
index c6c41055d6..786376c830 100644
--- a/builtin/logical/pki/storage.go
+++ b/builtin/logical/pki/storage.go
@@ -158,6 +158,10 @@ func importKey(ctx context.Context, s logical.Storage, keyValue string, keyName
        // and identifier); the last return field is whether or not an error
        // occurred.
        //
+       // Normalize whitespace before beginning. See note in importIssuer as to
+       // why we do this.
+       keyValue = strings.TrimSpace(keyValue) + "\n"
+
        // Before we can import a known key, we first need to know if the key
        // exists in storage already. This means iterating through all known
        // keys and comparing their private value against this value.
```